src/ngraph/node.hpp

```diff
@@ -188,7 +188,7 @@ namespace ngraph
         virtual bool supports_decompose() const { return false; }
         /// \brief Can replace a node with a constant during constant folding.
         /// \returns vector of outputs to replace node outputs. Empty outputs will not be replaced.
-        virtual OutputVector constant_fold() { return {}; }
+        virtual OutputVector constant_fold(const OutputVector& input_values) { return {}; }
```

Comment:
@diyessi, as we are moving away from using a local graph context for a node when evaluating value (completely or partially), I would suggest to rename this function to something like partial_infer or partial_evaluate or just evaluate (because partial becomes everywhere with more and more dynamic stuff involved). Probably a combination with `reference` word is also required to avoid confusion with real performance-oriented inference/evaluation in a HW plug-in.

Also I would suggest to keep an original method which will be a helper for case if we really want to evaluate a node based on the in-graph arguments. It will just call this new method with explicit `input_values` (as you already did later in this PR).

Reply:
`evaluate` seems fine to me. The actual "reference kernel" tends to be a separate template so that it can deal with all the needed element types in one place, so the `evaluate` method would be glue around instantiations of the reference kernel. That also allows a reference kernel to be used in more than one place, since it's just a template definition that executes on void*s.

Reply:
reference kernels are in general too slow to be used for constant folding. An example is the reference Broadcast kernel which took 114 seconds on BERT Large. Replacing the reference kernel with a more optimized kernel reduced this time to 2.5 seconds. Broadcast is a simple example, there are other reference kernels that can't be used in constant folding due to performance problems.

Reply:
We'll want to use the opt kernels when they exist. They fall back on reference kernels for the non-opt cases anyways.